cirq/protocols/apply_unitary.py

```diff
@@ -85,28 +89,122 @@ def __init__(self,
                 dtype as the target tensor.
             axes: Which axes the unitary effect is being applied to (e.g. the
                 qubits that the gate is operating on).
+
         """
         self.target_tensor = target_tensor
         self.available_buffer = available_buffer
         self.axes = tuple(axes)
+        self._is_subspace = _is_subspace
 
     @staticmethod
-    def default(num_qubits: int) -> 'ApplyUnitaryArgs':
-        """A default instance starting in state |0⟩."""
+    def default(num_qubits: Optional[int] = None,
+                *,
+                qid_shape: Optional[Tuple[int, ...]] = None
+               ) -> 'ApplyUnitaryArgs':
+        """A default instance starting in state |0⟩.
+
+        Specify exactly one argument.
+
+        Args:
+            num_qubits: The number of qubits to make space for in the state.
+            qid_shape: The shape of the state, specifying the dimension of each
+                qid."""
+        if (num_qubits is None) == (qid_shape is None):
+            raise TypeError(
+                'Specify either the num_qubits or qid_shape argument.')
+        if num_qubits is not None:
+            qid_shape = (2,) * num_qubits
+        assert qid_shape is not None, "Can't be None here"  # Satisfy mypy
+        num_qubits = len(qid_shape)
         state = linalg.one_hot(index=(0,) * num_qubits,
-                               shape=(2,) * num_qubits,
+                               shape=qid_shape,
                                dtype=np.complex128)
         return ApplyUnitaryArgs(state, np.empty_like(state), range(num_qubits))
 
-    def subspace_index(self, little_endian_bits_int: int
-                       ) -> Tuple[Union[slice, int, 'ellipsis'], ...]:
+    def with_transposed_and_sliced_tensors(
+            self, indices: Iterable[int],
+            slices: Iterable[Union[int, slice, Sequence[int]]]
```

Comment:
Make this argument optional and merge this method with the "for_operation_with_qid_shape" method.

Reply:
If I merge them, I'll loose some of the functionality of this method because the shape argument of `for_operation_with_qid_shape` is converted to `slices=[slice(0, s) for s in shape]`.

Reply:
Removed `with_transposed_and_sliced_tensors`.